src/App/Datums.cpp

```diff
@@ -21,42 +23,318 @@
+
+    auto* feature = static_cast<App::DatumElement*>(featureObj);
+    feature->Placement.setValue(Base::Placement(Base::Vector3d(), data.rot));
+    feature->Role.setValue(data.role);
+
+    feature->Placement.setStatus(Property::Hidden, true);
+    return feature;
+}
+
+LocalCoordinateSystem::SetupData LocalCoordinateSystem::getData(const char* role)
+{
+    const auto& setupData = getSetupData();
+    for (auto data : setupData) {
+        if (std::strcmp(role, data.role) == 0) {
+            return data;
+        }
+    }
+    return LocalCoordinateSystem::SetupData();
+}
+
+void LocalCoordinateSystem::setupObject()
+{
+    std::vector<App::DocumentObject*> links;
+    const auto& setupData = getSetupData();
+    for (auto data : setupData) {
+        links.push_back(createDatum(data));
+    }
+
+    OriginFeatures.setValues(links);
+}
+
+void LocalCoordinateSystem::unsetupObject()
+{
+    const auto& objsLnk = OriginFeatures.getValues();
+    // Copy to set to assert we won't call methode more then one time for each object
+    std::set<App::DocumentObject*> objs(objsLnk.begin(), objsLnk.end());
+    // Remove all controlled objects
+    for (auto obj : objs) {
+        // Check that previous deletes wasn't inderectly removed one of our objects
+        const auto& objsLnk = OriginFeatures.getValues();
+        if (std::find(objsLnk.begin(), objsLnk.end(), obj) != objsLnk.end()) {
+            if (!obj->isRemoving()) {
+                obj->getDocument()->removeObject(obj->getNameInDocument());
+            }
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+LocalCoordinateSystem::LCSExtension::LCSExtension(LocalCoordinateSystem* obj)
+    : obj(obj)
+{
+    Group.setStatus(Property::Transient, true);
+}
+
+void LocalCoordinateSystem::LCSExtension::initExtension(ExtensionContainer* obj)
+{
+    App::GroupExtension::initExtension(obj);
+}
+
+bool LocalCoordinateSystem::LCSExtension::extensionGetSubObject(DocumentObject*& ret,
+                                                                const char* subname,
+                                                                PyObject** pyobj,
+                                                                Base::Matrix4D* mat,
+                                                                bool,
+                                                                int depth) const
+{
+    if (!subname || subname[0] == '\0') {
+        return false;
+    }
+
+    // mapping of object name to role name
+    std::string name(subname);
+    for (int i = 0; i < 3; i++) {
+        if (name.rfind(LocalCoordinateSystem::AxisRoles[i], 0) == 0) {
+            name = LocalCoordinateSystem::AxisRoles[i];
+            break;
+        }
+        if (name.rfind(LocalCoordinateSystem::PlaneRoles[i], 0) == 0) {
+            name = LocalCoordinateSystem::PlaneRoles[i];
+            break;
+        }
+    }
```

Comment:
I know this is old code, but we should mark it for a refactor: this is pretty opaque.